### PR TITLE
Prepack conv weights

### PR DESCRIPTION
**Pull request by pinzhenx**

### CODE REVIEW

File: torch_ipex/csrc/aten_ipex_bridge.cpp [hunk 1]
```diff
@@ -105,6 +104,41 @@ void reorderDilTensorToPublic(const at::Tensor& ipexTensor) {
...
+  } else {
+    // If tensor is of blocked format, cpu raw data means nothing here.
+    new_shade_data_context->cpu_raw_data = nullptr;
+    new_shade_data_context->cpu_del_fun = nullptr;
+  }
+
+  // Create a new DataPtr instances because the DataPtr class does not support set
+  // its data or context directly
+  c10::DataPtr shade_data_ptr(
+    dst.get_data_handle(),
```
Comment (EikanWang) [inline torch_ipex/csrc/aten_ipex_bridge.cpp:131]: We cannot set the data handle of ideep tensor to Aten tensor data if ideep tensor is with block format. Suggest to use `new_shade_data_context->cpu_raw_data`

File: torch_ipex/csrc/cpu/DevOPs.cpp [hunk 2]
```diff
@@ -160,9 +164,27 @@ at::Tensor AtenIpexCPUDev::dil_convolution_overrideable(const at::Tensor & input
...
+              padding.vec(),
+              dilation.vec(),
+              groups,
+              dil::algorithm::convolution_direct);
+      bridge::reorderDilTensorGeneric(weight, packed_desc);
+    }
+
     return dil_convolution(
       input.is_contiguous() ? input : input.contiguous(),
-      weight.is_contiguous() ? weight : weight.contiguous(),
```
Comment (EikanWang) [inline torch_ipex/csrc/cpu/DevOPs.cpp:165]: We have an assumption: All the aten tensor should be contiguous if its buffer is the ideep tensor with block-format. So does the reordered weight tensor should always be contiguous?

File: torch_ipex/csrc/cpu/ShadeDataContext.h [hunk 3]
```diff
@@ -101,9 +101,6 @@ struct ShadeDataContext {
         TORCH_INTERNAL_ASSERT_DEBUG_ONLY(check_tensor_own_whole_storage(tensor));
         return true;
       } else {
-        // The dnnl tensor shares some data with raw tensor.
-        TORCH_INTERNAL_ASSERT_DEBUG_ONLY(shade_data_context->dil_tensor->is_public_format());
```
Comment (EikanWang) [inline torch_ipex/csrc/cpu/ShadeDataContext.h:105]: Did you hit the assert?
1. If the ideep tensor is block-format, then the data of aten tensor should be nullptr.
2. If the ideep tensor is plain-format, then the data of aten tensor should be as same as the data handle of the ideep tensor.

File: torch_ipex/csrc/cpu/ShadeDataContext.h [hunk 4]
```diff
@@ -116,7 +113,6 @@ struct ShadeDataContext {
         // All these tensors share same buffer of Tensor A with different storge offsets and elements.
         // So the context modification will impact all these tensors.
         if (check_tensor_own_whole_storage(tensor)) {
-          TORCH_INTERNAL_ASSERT_DEBUG_ONLY(shade_data_context->dil_tensor->get_size() == tensor.storage().capacity());
```
Comment (EikanWang) [inline torch_ipex/csrc/cpu/ShadeDataContext.h:119]: Did you hit the assert? Or is it a reduandant assert?

File: torch_ipex/csrc/aten_ipex_bridge.cpp (hunk 1 above)
Comment (pinzhenx) [reply torch_ipex/csrc/aten_ipex_bridge.cpp:131]: got it

File: torch_ipex/csrc/cpu/ShadeDataContext.h (hunk 3 above)
Comment (pinzhenx) [reply torch_ipex/csrc/cpu/ShadeDataContext.h:105]: it was caused by previous incorrect setting of aten tensor data. Once I fix the shade_data_ptr, the assertion failure was gone.

File: torch_ipex/csrc/cpu/ShadeDataContext.h (hunk 4 above)
Comment (pinzhenx) [reply torch_ipex/csrc/cpu/ShadeDataContext.h:119]: ditto

**Comment by jgong5** [comment]
I have a concern related but not specific to this PR w.r.t. the in-place changes of input data/weight tensors. PyTorch originally assumes the input tensors as constant and it is safe for re-entrance, e.g. called by multiple threads from JIT with fork and join. The in-place changes of these input tensors might break this assumption. Do we have the design to protect these in-place changes from multi-threaded access?

**Comment by pinzhenx** [comment]
As of right now, no. Modifying constant tensors on the fly, especially constant parameters, do have some implications we need to take care of.

**Comment by pinzhenx** [comment]
BTW, we are also implementing a standalone prepack jit pass. This PR is kind of a workaround to optimize the imperative path, so that we could achieve the same performance as the `to_mkldnn(model)` approach.

We may revert this design in the future if customers are more inclined to use JIT, or if something went wrong with it.

**Comment by jgong5** [comment]
Use cases for your consideration:
1. Shared weight by multiple in-process inference models in both imperative and JIT mode.
2. Shared data by multiple ops.
For 1), pre-packing during model initialization should work fine. But pre-packing on the first model run needs extra care on re-entrance.
For 2), there won't be re-entrance in the imperative path since there is just one model but maybe there is a problem if some downstream ops need blocked layout while others need plain layout. Also, we need to take care of the re-entrance in the JIT mode.

**Comment by EikanWang** [comment]
Thanks, Jiong. I discussed with Pinzhen, for case 1, we can capture `module.to` and then prepack the weight tensors.

**Comment by pinzhenx** [comment]
We currently adopt prepacking conv weights in `module.to`

### Pros:
- thread-safe
- more explicit, in terms of comparing to implicitly prepack at runtime

### Cons:
- no input info. meaning that queried format might not be optimal. This is not manifest in conv but this could be a problem for linear weights.
- unable to re-pack conv weights if it's been reordered back to plain (maybe even not a con)

**Comment by jgong5** [comment]
> no input info. meaning that queried format might not be optimal. This is not manifest in conv but this could be a problem for linear weights.

It will be a problem for conv too if we use winograd. Just FYI.

File: intel_pytorch_extension_py/ops/module.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+    def prepack_reccur(m):
+        prepack(m)
+        for _, sub_m in m.named_children():
+            prepack_reccur(sub_m)
+
+    m = orig_module_to(self, *args, **kwargs)
+
+    device = torch._C._nn._parse_to(*args, **kwargs)[0]
+    if device and device.type == 'dpcpp':
```
Comment (EikanWang) [inline intel_pytorch_extension_py/ops/module.py:20]: We need to check auto_dnnl here. If the user disables auto_dnnl, it should go through the original path.

**Pull request merged**